### PR TITLE
GCP support for configuring existing worker nodes as gw nodes

### PR DESCRIPTION
**Pull request by sridhargaddam**
This PR implements the necessary functionality to configure the
existing worker nodes of OCP cluster as Submariner Gateway nodes.
In this process, it mainly does the following.
1. Identifies worker nodes per zone and labels them with submariner
   gateway label.
2. Configures public-ip on the respective instance
3. Modifies the GCP instance NetworkTag to include the label used
   for external firewall configuration.

Fixes issue: https://github.com/submariner-io/cloud-prepare/issues/97
Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

### CODE REVIEW

**Comment by sridhargaddam** [comment]
CI License header check seems to be failing for unrelated issue.

**Review by skitt — APPROVED** [review]

**Review by aswinsuryan — APPROVED** [review]

File: pkg/gcp/gcp.go [hunk 1]
```diff
@@ -33,14 +33,11 @@ const (
 	messageRetrievedZones             = "Retrieved the zones"
 	messageValidateCurrentGWCount     = "Verifying if current gateways match the required number of gateways"
 	messageValidatedCurrentGWs        = "Current gateways match the required number of gateways"
-	messageDeployGatewayNode          = "Deploying gateway node"
 	messageDeployedGatewayNode        = "Successfully deployed gateway node"
 	messageInsufficientZonesForDeploy = "there are insufficient zone instances to deploy the required number of gateways"
 	messageCreateExtFWRules           = "Configuring the required firewall rules for inter-cluster traffic"
-	messageDeleteExtFWRules           = "Deleting the Submariner gateway firewall rules"
+	messageDeleteExtFWRules           = "Retrieving the Submariner gateway firewall rules"
```
Comment (tpantelis) [inline pkg/gcp/gcp.go]: Why was this changed? Now the text doesn't match the constant name.

File: pkg/gcp/ocpgwdeployer.go [hunk 2]
```diff
@@ -67,11 +79,87 @@ func (d *ocpGatewayDeployer) Deploy(input api.GatewayDeployInput, reporter api.R
...
+
+	// Currently, we only support increasing the number of Gateway nodes which could be a valid use-case
+	// to convert a non-HA deployment to an HA deployment. We are not supporting decreasing the Gateway
+	// nodes (for now) as it might impact the datapath if we accidentally delete the active GW node.
+	if numGatewayNodes < input.Gateways {
+		gatewayNodesToDeploy := input.Gateways - numGatewayNodes
+
+		if d.dedicatedGWNode {
+			for _, zone := range eligibleZonesForGW.Elements() {
+				reporter.Started(fmt.Sprintf("Deploying dedicated gateway node in zone %q", zone))
```
Comment (tpantelis) [inline pkg/gcp/ocpgwdeployer.go:101]: nit but we're using a mix of "gateway" and "Gateway" in the messages. I think we should be consistent. I prefer "gateway".

File: pkg/gcp/ocpgwdeployer.go [hunk 3]
```diff
@@ -249,17 +331,41 @@ func (d *ocpGatewayDeployer) Cleanup(reporter api.Reporter) error {
...
+			// the gateway node was deployed using the OCPMachineSet API otherwise it's an existing worker node.
 			prefix := d.gcp.infraID + "-submariner-gw-" + zone.Name
 			if strings.HasPrefix(instance.Name, prefix) {
+				reporter.Started(fmt.Sprintf("Deleting the Gateway instance %q", instance.Name))
 				err := d.deleteGateway(zone.Name)
+				if err != nil {
+					return reportFailure(reporter, err, "failed to delete dedicated gateway instance %q", instance.Name)
+				}
+
+				reporter.Succeeded("Successfully deleted the instance")
```
Comment (tpantelis) [inline pkg/gcp/ocpgwdeployer.go:348]: I notice we're using a mix of message string constants and inlining the strings. I think we should be consistent. I prefer inlining as you've done here.

**Review by tpantelis — CHANGES_REQUESTED** [review]
Looks good - just some nits.

File: pkg/gcp/gcp.go (hunk 1 above)
Comment (sridhargaddam) [reply pkg/gcp/gcp.go]: Sure, I'll rename the constant accordingly. Basically, I wanted the output to be as follows, so changed the text.

```
[sgaddam@localhost submariner-operator]$ ./bin/subctl cloud cleanup gcp --ocp-metadata ~/workspace-intellij/aws-ocp/sgaddam-
gcp-cp/metadata.json 
 ✓ Retrieving GCP credentials from your GCP configuration
 ✓ Initializing GCP connectivity
 ✓ Retrieving the Submariner gateway firewall rules
 ✓ Deleting firewall rule "sgaddam-gcp-cp-wrskd-submariner-public-ports-ingress" on GCP 
 ✓ Deleted firewall rule "sgaddam-gcp-cp-wrskd-submariner-public-ports-ingress" on GCP
 ✓ Successfully deleted the firewall rules
 ✓ Retrieving zones in the project 
...
```

File: pkg/gcp/ocpgwdeployer.go (hunk 2 above)
Comment (sridhargaddam) [reply pkg/gcp/ocpgwdeployer.go:101]: Done

File: pkg/gcp/ocpgwdeployer.go (hunk 3 above)
Comment (sridhargaddam) [reply pkg/gcp/ocpgwdeployer.go:348]: I was trying to follow the existing pattern used in the existing AWS code - https://github.com/submariner-io/cloud-prepare/blob/579b81efb86d10ed7c89bc46d6bd9ae4b1a72d85/pkg/aws/aws.go#L25

But even I had the same feeling. After this PR is merged, I'll push a separate PR and use the inline string. I hope you are okay if I do it in a separate PR.

**Review by sridhargaddam — COMMENTED** [review]
Thanks all for reviewing.

**Review by tpantelis — APPROVED** [review]
